### ISSUE TITLE
Feature and Enhancement Proposals after a review

### ISSUE THREAD

**Issue by mehmetcavus** [gocd/gocd#6805]
##### Issue Type

- Feature proposal
- Feature enhancement

After a month time of usage and review, sharing a feedback. If it worths escalating, I can create separate issues. Let me know which ones. 

##### Summary

1. Dependency Deadlocks and Resolution:
Renaming a pipeline, material, stage, job causes causes dependency deadlocks and it is a pain to resolve manually even for simple pipeline flows. Cannot rename pipeline/stage/job because it's a downstream dependency, or cannot rename a material because it's used in pipeline label template.

2. Config Repository Pipelines:
Users need to navigate to config repo for checking configuration of pipelines which means going through xml. This is also a pain because of xml interpretation, references etc.

3. Fetch Once and Propagate:
Having ability to fetch materials only once and propagate to other pipelines as defined. While using identical or very similar environments with different deployment configurations -as in the best practices-, fetching again and again is just wasting resources. Currently using scripts as a workaround. 

4. SCM Revision Availability:
SCM Revision (GO_REVISION) of Material M1 defined in Pipeline P1 is not accessible in Pipeline PN. Defining it as a material in Pipeline PN and unselecting stage level "Fetch Materials" option effects other materials defined in the pipeline too.

5. Global & Top Level Parameters, Environment Variables, Secure Variables:
Since GoCD does not have project concept, container of a pipeline flow is either a group or an environment. To define once and have access to e.g. ${PRIVATE_DOCKER_REGISTRY_HOST} in all pipelines requires creating an environment and assigning these pipelines to this environment, relevant or not. Users are enforced to this at the moment as a workaround. This has side effects like handling resource assignment (agents). To avoid this, it can be defined in pipeline level but then managing the change is another hurdle. Too much for too less.

6. Renaming, Deleting Environments:
This requires editing Config XML. 

7. Changing Task Orders:
Changing task order one position at a time not user friendly.

8. Automating fresh installation with configurations, but without Config Repository:
Ability to automate setting up _initial desired state of a fresh copy of GoCD environment with configurations (including plugins and their configuration)_ via tools like Ansible. This will allow users to use tools like Ansible without any previous installation / usage of GoCD. Currently, it's only possible by using Config Repository but this is not a fresh and dynamic install. Also, any kind of key should not be stored in SCM including agentAutoRegisterKey, tokenGenerationKey.

###### Basic environment details

* Go Version: `19.3.0, 19.5.0, 19.7.0`
* JAVA Version: `12.0.1`
* OS: `Ubuntu 18.04.02, Linux 4.9.184-linuxkit`

##### Expected Results

1. Dependency Deadlocks and Resolution: 
Dependency Deadlocks should be resolved automatically. User should have a preference for dependency resolution in global, group and environment levels: "Ask/Prompt me" or "Resolve automatically". If a situation cannot be resolved automatically, it should prompt user with available resolution options including a suggestion for manual intervention, even if the preference is "Resolve automatically".

2. Config Repository Pipelines:
Read only view of config repository pipelines should be possible with a preference on pipeline and config repository level: Enable/Disable read-only view. The gear icon is used for opening pipeline settings which doesn't always mean that there will be an edit. Setting at config repo level > Setting at pipeline level

3. Fetch Once and Propagate:
Git Materials can be defined like how Pipeline Materials defined now, and Pipeline PN can have access materials of its downstream pipeline P1, so properties of a material too. It can be used like ${git_material_name.GO_REVISION}. So, they can be used in deployment scripts etc.

4. SCM Revision Availability:
Item 3 resolves this too.

5. Global & Top Level Parameters, Environment Variables, Secure Variables:
Adding ability to define them in GoCD Level, Group Level and Environment Level. Each helps different use-cases. If there is GoCD Level Environment Variable, this means it will be available to all pipelines, like ${PRIVATE_DOCKER_REGISTRY_HOST}. 

6. Renaming, Deleting Environments:
It should be possible to delete, rename an environment. 

7. Changing Task Orders:
It should be possible to drag and drop to change order of a task.

8. Automating fresh installation with configurations, but without Config Repository:
This should not require GoCD API usage, because it's automation and a fresh install. By just providing installation configuration file, initial desired state of fresh GoCD environment can be automated in this way. This means agentAutoRegisterKey, tokenGenerationKey generated during go-server installation should be injected to go-agent installation (inputs and outputs of an installation). Dockerfile and docker-compose files should be supported.

##### Any other info

Thanks in advance for your comments.

**Comment by adityasood** [comment]
@mehmetcavus some of my thoughts on your suggestions

> Dependency Deadlocks and Resolution:
> Renaming a pipeline, material, stage, job causes causes dependency deadlocks and it is a pain to resolve manually even for simple pipeline flows. Cannot rename pipeline/stage/job because it's a downstream dependency, or cannot rename a material because it's used in pipeline label template.

Yes currently renaming these entities is not allowed within GoCD. I agree it might help to have that option. There are a lot of questions for e.g what happens when a pipeline name is changed and then it is re-used by another pipeline. Currently, pipeline history is also dependent on the pipeline name.

We do have some suggestions being discussed, for e.g we have an open PR around [Add display names to pipelines
](https://github.com/gocd/gocd/pull/3183)

Or adding a display name and description field: https://github.com/gocd/gocd/issues/6366

> Config Repository Pipelines:
> Users need to navigate to config repo for checking configuration of pipelines which means going through xml. This is also a pain because of xml interpretation, references etc.

We have quite a few enhancements being currently worked upon to improve Config Repo experience.

The stream of work is under the label: https://github.com/gocd/gocd/issues?q=is%3Aissue+is%3Aopen+label%3A%22Pipelines+as+Code%22

For e.g: We have started listing all pipelines which are defined in a config repo on the config repo configuration and it is a clickable link which then takes you to the pipeline run history.

We are also working on a UI/UX refresh of pipeline config page and are planning to include a read only view for config repo pipelines as part of this work.

Link to workstream: https://github.com/gocd/gocd/labels/pipeline-config-spa


> Fetch Once and Propagate:
> Having ability to fetch materials only once and propagate to other pipelines as defined. While using identical or very similar environments with different deployment configurations -as in the best practices-, fetching again and again is just wasting resources. Currently using scripts as a workaround.
> 
> SCM Revision Availability:
> SCM Revision (GO_REVISION) of Material M1 defined in Pipeline P1 is not accessible in Pipeline PN. Defining it as a material in Pipeline PN and unselecting stage level "Fetch Materials" option effects other materials defined in the pipeline too.

Could you help me understand what do you mean by propogation? Is your use case around making the material revision of the upstream pipelines available for downstream pipelines as envrionment variable?

You mentioned you currently use scripts to do this, could you explain what the scripts do?

> Global & Top Level Parameters, Environment Variables, Secure Variables:
> Since GoCD does not have project concept, container of a pipeline flow is either a group or an environment. To define once and have access to e.g. ${PRIVATE_DOCKER_REGISTRY_HOST} in all pipelines requires creating an environment and assigning these pipelines to this environment, relevant or not. Users are enforced to this at the moment as a workaround. This has side effects like handling resource assignment (agents). To avoid this, it can be defined in pipeline level but then managing the change is another hurdle. Too much for too less.

The closest this can be done in GoCD is by using the Secrets Management to define these variables externally. You can read more here: https://docs.gocd.org/current/configuration/secrets_management.html.

> Renaming, Deleting Environments:
> This requires editing Config XML.

 That's right. There's no way to delete environments using the UI. Our usual recommendation is to use the API https://api.gocd.org/19.8.0/#environment-config

> Changing Task Orders:
> Changing task order one position at a time not user friendly.

We are working on revamping the Pipeline Config Page. This should be handled in this stream of work. 

GitHub link to work stream: https://github.com/gocd/gocd/issues?q=is%3Aissue+label%3Apipeline-config-spa

> Automating fresh installation with configurations, but without Config Repository:
> Ability to automate setting up initial desired state of a fresh copy of GoCD environment with configurations (including plugins and their configuration) via tools like Ansible. This will allow users to use tools like Ansible without any previous installation / usage of GoCD. Currently, it's only possible by using Config Repository but this is not a fresh and dynamic install. Also, any kind of key should not be stored in SCM including agentAutoRegisterKey, tokenGenerationKey.

For your use case you can do something like below. GoCD can come up with a known configuration. 

```
mkdir -p config/ plugins/external
# populate config/cruise-config.xml (maybe using [gomatic](https://github.com/gocd-contrib/gomatic))
# add plugins into plugins/external
```

We have users who have used Ansible and Chef cookbooks to automate this at their end. GoCD binary is available in all forms as docker images, OS specific distributions and helm chart. With tools like Ansible it should also be possible to templatize the config XML itself, including the keys.

We have elastic agents for automated setups where users can bring up agents on demand and do not have to worry about configuring autor-register properties explicitly.

Link to GoCD Docker Images: https://hub.docker.com/search?q=gocd&type=image

**Comment by mehmetcavus** [comment]
> @mehmetcavus some of my thoughts on your suggestions
> 
> > Dependency Deadlocks and Resolution:
> > Renaming a pipeline, material, stage, job causes causes dependency deadlocks and it is a pain to resolve manually even for simple pipeline flows. Cannot rename pipeline/stage/job because it's a downstream dependency, or cannot rename a material because it's used in pipeline label template.
> 
> Yes currently renaming these entities is not allowed within GoCD. I agree it might help to have that option. There are a lot of questions for e.g what happens when a pipeline name is changed and then it is re-used by another pipeline. Currently, pipeline history is also dependent on the pipeline name.

As long as adding display name works for all dependency locks (pipeline, material, stage, job, what else?), #6366 is fine I think. Hope it is extended in this way and added to a release very soon. Because this seems serious showstopper for the future, and gets bigger. For example: while there is pipeline "pipe1", I added a config repo (having another pipeline "pipe1") which caused a conflict saying:
```
INVALID MERGED CONFIGURATION
Number of errors: 8+
1. You have defined multiple pipelines named 'pipe1'. Pipeline names must be unique.  …
…
```
To be able to add config repo, resolving 8+ errors in existing pipe1 because of dependency deadlocks... Also, the issue breaks automated processes.
.
> 
> We do have some suggestions being discussed, for e.g this PR around [Add display names to pipelines ](https://github.com/gocd/gocd/pull/3183)
> 
> Or adding a display name and description field: #6366
> 
> > Config Repository Pipelines:
> > Users need to navigate to config repo for checking configuration of pipelines which means going through xml. This is also a pain because of xml interpretation, references etc.
> 
> We have quite a few enhancements being currently worked upon to improve Config Repo experience.
> 
> The stream of work is under the label: https://github.com/gocd/gocd/issues?q=is%3Aissue+is%3Aopen+label%3A%22Pipelines+as+Code%22
> 
> For e.g: We have started listing all pipelines which are defined in a config repo on the config repo configuration and it is a clickable link which then takes you to the pipeline run history.
> 
> We are also working on a UI/UX refresh of pipeline config page and are planning to include a read only view for config repo pipelines as part of this work.
> 
> Link to workstream: https://github.com/gocd/gocd/labels/pipeline-config-spa

I see that read-only view added to the project Pipeline Config UX, good news! Thanks.
.
> 
> > Fetch Once and Propagate:
> > Having ability to fetch materials only once and propagate to other pipelines as defined. While using identical or very similar environments with different deployment configurations -as in the best practices-, fetching again and again is just wasting resources. Currently using scripts as a workaround.
> > SCM Revision Availability:
> > SCM Revision (GO_REVISION) of Material M1 defined in Pipeline P1 is not accessible in Pipeline PN. Defining it as a material in Pipeline PN and unselecting stage level "Fetch Materials" option effects other materials defined in the pipeline too.
> 
> Could you help me understand what do you mean by propogation? Is your use case around making the material revision of the upstream pipelines available for downstream pipelines as envrionment variable?
> 
> You mentioned you currently use scripts to do this, could you explain what the scripts do?

I mean “transfer / make available to next phases” here. And yes, as environment variable. 

**Scenario**:
1. In pipeline P1, a nodejs repo is fetched from github as material M1 and has another material M2. (let’s call P1.M1 and P1.M2 here)
2. Pipeline “deploy-to-dev” has same repo as M1. (deploy-to-dev.M1)
3. Pipeline “deploy-to-prod” has same repo as M1. (deploy-to-prod.M1)
4. Pipeline PN needs $GO_REVISION of P1.M1 (can be needed in scripts, or in pipeline label template etc.)

**Issue A**: Same repo, same commit to go through CI/CD pipelines. And steps above means fetching the repo from github 3 times when “Fetch Materials” checkbox selected on Pipeline Stage settings. 

**Issue B**: PN has to add the same github repo as material to access its $GO_REVISION, mentioned in step 4. This means fetching the repo from github once more. 

**Solution steps**: (for both issue)
1. Materials will be kept like artifacts when the new material setting “make available in downstream pipelines” is checked.
2. Downstream pipelines will add materials from upstream pipelines in the same way adding Pipeline as Material (Click “Add Material” button, select “Pipeline”, …). E.g. selecting “Upstream Material” or “Existing Material”.
3. Enable pipeline PN to access properties of material P1.M1 as environment variable.

About what I do with scripts (it was unclear):
In pipeline P1:
1. Generate a file having info e.g. $GO_REVISION of material P1.M1
2. tar-zip repo files of P1.M1 including node_modules folder (after running npm install) 
3. copy the zip file to “dist” directory under pipeline working directory
4. Add build artifact. Source: “dist” folder, Target: dist folder

In pipeline deploy-to-dev, deploy-to-prod: (“Fetch Materials” unchecked)

5. Fetch artifact from pipeline P1 (source dir: “dest”)
6. untar files to working directory
7. access $GO_REVISION of material P1.M1 from the file generated in step 1
8. … (finally start things for this pipeline regarding its environment e.g. dev, prod)

You can ignore this part (extra info): 
Why build Artifacts not used for repo files of P1.M1 as well, but using tar-zip instead: Mounted /godata folder, so want to make sure that files are not changed in local environment by any chance between the subsequent pipeline runs during development of GoCD workflow and deployment scripts. (When P2 runs repeatedly, it takes .tar file from latest run of P1 even though files in P1 working directory might have been changed later in local (uncompleted code)).
.
> 
> > Global & Top Level Parameters, Environment Variables, Secure Variables:
> > Since GoCD does not have project concept, container of a pipeline flow is either a group or an environment. To define once and have access to e.g. ${PRIVATE_DOCKER_REGISTRY_HOST} in all pipelines requires creating an environment and assigning these pipelines to this environment, relevant or not. Users are enforced to this at the moment as a workaround. This has side effects like handling resource assignment (agents). To avoid this, it can be defined in pipeline level but then managing the change is another hurdle. Too much for too less.
> 
> The closest this can be done in GoCD is by using the Secrets Management to define these variables externally. You can read more here: https://docs.gocd.org/current/configuration/secrets_management.html.

Only GoCD file-based secret plugins is available. Looking forward to a nice way to integrate with Hashicorp Vault until a plugin is released.
.
> 
> > Renaming, Deleting Environments:
> > This requires editing Config XML.
> 
> That's right. There's no way to delete environments using the UI. Our usual recommendation is to use the API https://api.gocd.org/19.8.0/#environment-config

If there is, what are the side-affects for renaming from cr­uise-­confi­g.xml? (There might be a reason why API usage is recommended for this)
.
> 
> > Changing Task Orders:
> > Changing task order one position at a time not user friendly.
> 
> We are working on revamping the Pipeline Config Page. This should be handled in this stream of work.
> 
> GitHub link to work stream: https://github.com/gocd/gocd/issues?q=is%3Aissue+label%3Apipeline-config-spa
Ok, will join discussion there if needed.
> 
> > Automating fresh installation with configurations, but without Config Repository:
> > Ability to automate setting up initial desired state of a fresh copy of GoCD environment with configurations (including plugins and their configuration) via tools like Ansible. This will allow users to use tools like Ansible without any previous installation / usage of GoCD. Currently, it's only possible by using Config Repository but this is not a fresh and dynamic install. Also, any kind of key should not be stored in SCM including agentAutoRegisterKey, tokenGenerationKey.
> 
> For your use case you can do something like below. GoCD can come up with a known configuration.
> 
> ```
> mkdir -p config/ plugins/external
> # populate config/cruise-config.xml (maybe using [gomatic](https://github.com/gocd-contrib/gomatic))
> # add plugins into plugins/external
> ```
> 
> We have users who have used Ansible and Chef cookbooks to automate this at their end. GoCD binary is available in all forms as docker images, OS specific distributions and helm chart. With tools like Ansible it should also be possible to templatize the config XML itself, including the keys.
> 
> We have elastic agents for automated setups where users can bring up agents on demand and do not have to worry about configuring autor-register properties explicitly.
> 
> Link to GoCD Docker Images: https://hub.docker.com/search?q=gocd&type=image

Already using GoCD server and agent docker images and custom scripts in https://hub.docker.com/r/gocd/gocd-server. Templating config file requires maintenance for each GoCD release. 

So, what I was thinking is that it should be possible to set server properties “agentAutoRegisterKey, webhookSecret, serverId, tokenGenerationKey” with docker commands like: 
```
docker run -e GOCD_TOKEN_GENERATION_KEY=<value> gocd/gocd-server:v19.7.0
```

Is there a logic for generation of these values which might be used for validation later on for any purpose? Are users allowed to use dummy values for this automation?

If there is no logic and we can use any value, than being able to set these variables with environment variables makes sense and is enough. So users can use them as input to configurations etc.

**Comment by ketan** [comment]
> So, what I was thinking is that it should be possible to set server properties “agentAutoRegisterKey, webhookSecret, serverId, tokenGenerationKey” with docker commands like:

These keys are auto generated using a [random UUID](https://docs.oracle.com/javase/7/docs/api/java/util/UUID.html#randomUUID()). So every server should have different values. If you're looking to change these values, they are perfectly safe to change. We recommend that you shut down the server and change the value in the `cruise-config.xml` file and start the server.

As for what these keys are used for:

- `agentAutoRegisterKey`: it is a shared secret used by agents (and elastic agents) when they need to [auto-register](https://docs.gocd.org/current/advanced_usage/agent_auto_register.html) with the gocd server.
- `webhookSecret`: used for [webhooks](https://api.gocd.org/current/#hosting-your-repository-on-bitbucket) delivered by github/bitbucket etc.
- `tokenGenerationKey`: secret used by the server to generate API tokens for agents
- `serverId`: this one is not a secret, but just a server identifier. This is used internally by elastic agent plugins, so that 2 different gocd servers using elastic agnets connecting to the same shared cloud providers can each identify their own agents

**Comment by adityasood** [comment]
@mehmetcavus thank you for explaining the use cases I understand them now.

For Fetch once and propagate, a way to achieve this would be to write an artifact plugin to fetch the env and return them as metadata for the fetch task. You might want to take a look at the docs [here](https://plugin-api.gocd.org/19.8.0/artifacts/#requests-from-the-gocd-server).

> Only GoCD file-based secret plugins is available. Looking forward to a nice way to integrate with Hashicorp Vault until a plugin is released. Any roadmap for this plugin?

We will be releasing the vault plugin shortly.

> If there is, what are the side-affects for renaming from cr­uise-­confi­g.xml? (There might be a reason why API usage is recommended for this)

There are no side-effects per se. It’s usually not recommended to directly edit the config XML to update/delete any GoCD entity. Config structure might change across GoCD releases whereas the behaviour of the GoCD API doesn’t change for a particular API version.

**Comment by kritika-singh3** [comment]
@mehmetcavus Few updates:
 - The read-only pipeline configuration for the pipelines from config-repo is now live - released in [GoCD 19.12.0](https://www.gocd.org/releases/#latest).
 - We have also revamped the Environments SPA. The new SPA allows for environment deletion. This feature is available from GoCD 19.11.0
 - The GoCD secrets plugin for HashiCorp's Vault has been released. Download link: https://github.com/gocd/gocd-vault-secret-plugin/releases/latest

Please do update and provide your feedback.
